### ISSUE TITLE
Failed design time build on changing TargetFramework

### ISSUE THREAD

**Issue by mavasani** [dotnet/roslyn-project-system#1479]
1. Create new .NET Standard library or .NetCore App.
2. Right-click project and choose Edit csproj and change `<TargetFramework>netcoreapp1.0</TargetFramework>` to `<TargetFramework>net451</TargetFramework>` and save. 
3. Change target framework from `<TargetFramework>net451</TargetFramework>` to `<TargetFramework>netcoreapp1.0</TargetFramework>` and save.

Got: Failed design time build in the Debug output window, which seems to be executed with a project snapshot using the prior target framework (net451), but the imported targets are as per the new target framework (netcoreapp1.0), and hence causes missing target error.

```
Target "DesignTimeXamlMarkupCompilation" in file "C:\Windows\Microsoft.NET\Framework\v4.0.30319\Microsoft.Xaml.targets" from project "C:\Users\mavasani\documents\visual studio 2017\Projects\ClassLibrary6\ClassLibrary6\ClassLibrary6.csproj" (target "CoreCompile" depends on it):
Done building target "DesignTimeXamlMarkupCompilation" in project "ClassLibrary6.csproj".
The target "DesignTimeMarkupCompilation" does not exist in the project.
Done building project "ClassLibrary6.csproj" -- FAILED.
Build FAILED.
```

I set a breakpoint in CPS `ProjectBuildSnapshotService.UpdateSnapshotCoreAsync` and see that the design time build is for TargetFramework net451, CoreCompileDependsOn property includes target `DesignTimeMarkupCompilation`, which exists in Microsoft.WinFx.Targets, which is only imported for NetFx or Silverlight projects. So I feel we get this error because we are trying to get the target to execute using TargetFramework netcoreapp1.0, causing us to not find this target.

**Comment by mavasani** [comment]
So the failing design time build on TargetFramework change seems to be similar to the underlying CPS issue with unit test discovery on TargetFramework change.. when the TF is changed, a design time build is immediately kicked off and CPS uses the project snaphot and properties from before the edit, but when it tries to do the DT build, certain targets files are missing as it uses the new project file to compute the targets to execute.
 
For example, if TargetFramework = net451, includes Microsoft.WinFx.Targets, which sets CoreCompile to execute a target `DesignTimeMarkupCompilation` defined in this target file. When TargetFramework is changed to netcoreapp1.0, CoreCompileDependsOn still has this target, but design time build never imports Microsoft.WinFx.Targets for netcoreapp1.0. and we can a failed design time build with the following error:

```
Target "DesignTimeXamlMarkupCompilation" in file "C:\Windows\Microsoft.NET\Framework\v4.0.30319\Microsoft.Xaml.targets" from project "C:\Users\mavasani\documents\visual studio 2017\Projects\ClassLibrary6\ClassLibrary6\ClassLibrary6.csproj" (target "CoreCompile" depends on it):
Done building target "DesignTimeXamlMarkupCompilation" in project "ClassLibrary6.csproj".
The target "DesignTimeMarkupCompilation" does not exist in the project.
Done building project "ClassLibrary6.csproj" -- FAILED.
Build FAILED.
```
 
Basically design time builds will break as soon you edit a property that changes the included targets, until the CPS project snapshot service catches up and generates a new project snapshot and kicks off a subsequent design time build.

Tagging @lifengl @jviau - is it possible to avoid the design time build on stale project snapshot? Otherswise, I recommend we make the targeted fix for TF change for RTM: https://github.com/dotnet/roslyn-project-system/pull/1478

**Comment by mavasani** [comment]
@lifengl @jviau Did you guys get a chance to look at this issue? I am unable to make further progress into debugging CPS design time builds.
Note that the failed design time build is causing lot of IDE weirdness (broken intellisense, squiggles, etc.) and can seemingly happen when user edits any property that changes the included targets. This can cause lot of subtle IDE issues.

**Comment by jviau** [comment]
@mavasani I am looking into it now.

**Comment by jviau** [comment]
The issue here is that CPS runs design time builds out-of-proc. In that state msbuild has auto-reload from disk enabled. This means when you quickly change from `netcoreapp1.0` to `net451` and back to `netcoreapp1.0`, CPS will pass the correct project instance with `net451` to msbuild , but it will auto-reload the new `netcoreapp1.0` project form disk and fail the build. Msbuild reloads from disk before CPS has reacted to the changing project file at all, so we can't stop the failed design time build. 

Eventually a successful design time build is processed as we react to the project file change, so why does intellisense and test explorer remain broken? Is down chain dataflow faulting? The `IProjectBuildSnapshot` is of interesting form - since the design time build outright never ran, there is no results for us to present. `TargetOutputs` and `TargetSuccess` will be empty dictionaries.

**Comment by davkean** [comment]
But who's figuring out what targets to execute? Builds should be consistent, I should never attempt to build targets that don't come from the project that was just loaded from disk.

**Comment by jviau** [comment]
I believe MSBuild is doing a combination of using properties from the instance we supply, but targets from the reloaded file on disk. 
The targets we are executing, which is part of the build submission:
```
CollectResolvedSDKReferencesDesignTime;
CollectPackageReferences;
ResolveComReferencesDesignTime;
ResolveProjectReferencesDesignTime;
ResolveAssemblyReferencesDesignTime;
CollectSDKReferencesDesignTime;
ResolvePackageDependenciesDesignTime;
CompileDesignTime;
```
The target that fails, `DesignTimeMarkupCompilation`, is not part of that set at all. CPS never asks for that soon-to-be-invalid target to build. MSBuild must be getting it from the `CoreCompileDependsOn` property from the project instance we supply. (The only 2 things we pass to msbuild in this case is the `ProjectInstance` and the list of targets).

**Comment by mavasani** [comment]
Ported to VSO [385643](https://devdiv.visualstudio.com/DevDiv/_workitems?id=385643&fullScreen=false&_a=edit) - this is CPS or MSBuild bug.
